kong/runloop/handler.lua

```diff
@@ -365,6 +372,190 @@ local function register_events()
 end
 
 
+local function get_rebuild_timeout()
+  if kong.configuration.database == "cassandra" then
+    return kong.configuration.cassandra_timeout / 1000
+  end
+
+  if kong.configuration.database == "postgres" then
+    return kong.configuration.pg_timeout / 1000
+  end
+
+  return 60
+end
+
+
+local function get_version(name)
+  if not kong.cache then
+    return "init"
+  end
+
+  local version, err = kong.cache:get(name .. ":version", TTL_ZERO, utils.uuid)
+  if err then
+    return nil, "could not ensure " .. name .. " is up to date: " .. err
```

Comment:
I see you changed this (internal) function to `return nil, err` which leads the calling code to handle the error (which is going to repeat same code all over?)

Reply:
Using this to handle the "returning errors instead of logging inside the function" here.

As I see it, the purpose of this function is a single one: to trigger an update (which can be async or not). The rest (version checking, semaphore acquiring) are preparatory steps, needed to accomplish that. If one of them fails, then the function should make that manifest by erroring.

If `rebuild` itself starts deciding what to do with errors, we get into this situation in which the caller handles some of the errors, and `rebuild` handles others. I'd rather have one or the other do it, but not both.

For now I have decided to let the caller handle errors. It can just log the error and continue (like we do in `get_updated_plugins_iterator` now). Or it can decide to ignore one particular kind of error, like we do in timer-based rebuilds, where we ignore `"timeout"` errors which occur when attempting to get the lock.